# ML-09 — Validation and Research Claim Audit

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/shaikhabdullahwaseem17-byte/vigilant-meme/blob/main/work/notebooks/w06_validation_audit.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Two paper findings + my methodology questions

*Pick two findings from the FlyRank research paper. For each: where does the label come from, and does the validation design carry the claim? Constructive tone.*

Cloning repo...
Cloning into '/content/vigilant-meme'...
remote: Enumerating objects: 162, done.
remote: Counting objects: 100% (162/162), done.
remote: Compressing objects: 100% (113/113), done.
remote: Total 162 (delta 66), reused 98 (delta 31), pack-reused 0 (from 0)
Receiving objects: 100% (162/162), 1.87 MiB | 12.27 MiB/s, done.
Resolving deltas: 100% (66/66), done.
CWD: /content/vigilant-meme
sys.path[0]: /content/vigilant-meme
Imported scripts.ml_utils OK
PROCESSED_DIR: /content/vigilant-meme/data/processed


 **Label origin**: `is_declining_label` is derived from `trend_direction`, which itself is computed from `trend_pct` in the raw CSV (the repo's feature_metadata.json confirms this).

- **Validation design used previously**: Week-5 used a stratified random split; an honest validation should also show grouped-by-client and time-aware splits because content and clients repeat.

- **Leakage risk:** `trend_direction` and `trend_pct` should never be used as features (they are label-derived). Check the feature set and exclude any `trend_*` columns.



- **Base rate**: always print df['is_declining_label'].mean() next to precision@K and AUC when reporting.






### ***Findings***

The transparent baseline ranks items by visibility, freshness risk, position opportunity and depth gap (weights: 0.40 / 0.30 / 0.25 / 0.05).
The target (is_declining_label) is derived from trend_direction / trend_pct in the raw CSV, so any trend_* or label-derived fields must be excluded from features to avoid leakage.


# **Methodology questions to answer here**

Where does the label come from (column and formula) and what is the label window?
Which split did I use originally (random/grouped/time-aware)? Does that represent deployment?
Which columns were excluded to prevent leakage (explicit list)?
Is the base rate printed next to every reported metric?

In [8]:
# 1) Two findings + methodology questions (numbers/checks)
import json
from pathlib import Path
import textwrap

ROOT = Path('.')
PROCESSED = ROOT / 'data' / 'processed'
OUT = ROOT / 'outputs'

# helper to load JSON if present
def load_json(p):
    if p.exists():
        return json.loads(p.read_text())
    return None

feature_meta = load_json(PROCESSED / 'feature_metadata.json')
baseline_meta = load_json(PROCESSED / 'baseline_metadata.json')
model_report = (OUT / 'model_report.md').read_text() if (OUT / 'model_report.md').exists() else None

print("Files found:")
print(" - feature_metadata.json:", bool(feature_meta))
print(" - baseline_metadata.json:", bool(baseline_meta))
print(" - outputs/model_report.md:", bool(model_report))
print()

# Produce two candidate findings (use these as starting sentences in the markdown)
findings = []
if model_report:
    # pick first two non-empty lines as candidate findings
    for line in model_report.splitlines():
        line = line.strip()
        if line:
            findings.append(line)
        if len(findings) >= 2:
            break
else:
    # fallback using baseline/model metadata
    if baseline_meta:
        findings.append(f"Baseline top score = {baseline_meta.get('top_score'):.3f}, median = {baseline_meta.get('median_score'):.3f}.")
        declining_rate = baseline_meta.get('declining_rate_top_50', None)
        if declining_rate is not None:
            findings.append(f"Top-50 declining rate (baseline): {declining_rate:.2%} (from baseline metadata).")
    else:
        findings.append("Finding A: The baseline scoring rule ranks pages by visibility, freshness risk, position opportunity, and depth gap (weights: 0.40/0.30/0.25/0.05).")
        findings.append("Finding B: The task target is 'is_declining_label' derived from trend_direction (a directional label).")

print("Candidate findings (edit into the markdown above):")
for i, f in enumerate(findings, 1):
    print(f"{i}. {f}")

print("\nMethodology questions to answer for each finding (copy into the notebook above):")
questions = [
    "Where does the label come from? (column, calculation, and its window — e.g. trend_direction computed from trend_pct over last 30 days)",
    "Does the validation split used in Week-5 mimic deployment? (random vs grouped vs time-aware — which did you use?)",
    "Could any candidate features be label-derived or overlap the label window (future leakage)?",
    "Is the base rate printed next to the metric used to claim performance?"
]
for q in questions:
    print("- " + q)

# short note for the markdown: recommended short answers to paste
note = textwrap.dedent("""
Short recommended answers (paste and adapt):
- Label origin: `is_declining_label` is derived from `trend_direction`, which itself is computed from `trend_pct` in the raw CSV (the repo's feature_metadata.json confirms this).
- Validation design used previously: Week-5 used a stratified random split; an honest validation should also show grouped-by-client and time-aware splits because content and clients repeat.
- Leakage risk: `trend_direction` and `trend_pct` should never be used as features (they are label-derived). Check the feature set and exclude any `trend_*` columns.
- Base rate: always print df['is_declining_label'].mean() next to precision@K and AUC when reporting.
""")
print("\n---\n", note)# 1) Two findings + methodology questions (numbers/checks)
import json
from pathlib import Path
import textwrap

ROOT = Path('.')
PROCESSED = ROOT / 'data' / 'processed'
OUT = ROOT / 'outputs'

# helper to load JSON if present
def load_json(p):
    if p.exists():
        return json.loads(p.read_text())
    return None

feature_meta = load_json(PROCESSED / 'feature_metadata.json')
baseline_meta = load_json(PROCESSED / 'baseline_metadata.json')
model_report = (OUT / 'model_report.md').read_text() if (OUT / 'model_report.md').exists() else None

print("Files found:")
print(" - feature_metadata.json:", bool(feature_meta))
print(" - baseline_metadata.json:", bool(baseline_meta))
print(" - outputs/model_report.md:", bool(model_report))
print()

# Produce two candidate findings (use these as starting sentences in the markdown)
findings = []
if model_report:
    # pick first two non-empty lines as candidate findings
    for line in model_report.splitlines():
        line = line.strip()
        if line:
            findings.append(line)
        if len(findings) >= 2:
            break
else:
    # fallback using baseline/model metadata
    if baseline_meta:
        findings.append(f"Baseline top score = {baseline_meta.get('top_score'):.3f}, median = {baseline_meta.get('median_score'):.3f}.")
        declining_rate = baseline_meta.get('declining_rate_top_50', None)
        if declining_rate is not None:
            findings.append(f"Top-50 declining rate (baseline): {declining_rate:.2%} (from baseline metadata).")
    else:
        findings.append("Finding A: The baseline scoring rule ranks pages by visibility, freshness risk, position opportunity, and depth gap (weights: 0.40/0.30/0.25/0.05).")
        findings.append("Finding B: The task target is 'is_declining_label' derived from trend_direction (a directional label).")

print("Candidate findings (edit into the markdown above):")
for i, f in enumerate(findings, 1):
    print(f"{i}. {f}")

print("\nMethodology questions to answer for each finding (copy into the notebook above):")
questions = [
    "Where does the label come from? (column, calculation, and its window — e.g. trend_direction computed from trend_pct over last 30 days)",
    "Does the validation split used in Week-5 mimic deployment? (random vs grouped vs time-aware — which did you use?)",
    "Could any candidate features be label-derived or overlap the label window (future leakage)?",
    "Is the base rate printed next to the metric used to claim performance?"
]
for q in questions:
    print("- " + q)

# short note for the markdown: recommended short answers to paste
note = textwrap.dedent("""
Short recommended answers (paste and adapt):
- Label origin: `is_declining_label` is derived from `trend_direction`, which itself is computed from `trend_pct` in the raw CSV (the repo's feature_metadata.json confirms this).
- Validation design used previously: Week-5 used a stratified random split; an honest validation should also show grouped-by-client and time-aware splits because content and clients repeat.
- Leakage risk: `trend_direction` and `trend_pct` should never be used as features (they are label-derived). Check the feature set and exclude any `trend_*` columns.
- Base rate: always print df['is_declining_label'].mean() next to precision@K and AUC when reporting.
""")
print("\n---\n", note)

Files found:
 - feature_metadata.json: True
 - baseline_metadata.json: False
 - outputs/model_report.md: True

Candidate findings (edit into the markdown above):
1. # FlyRank Refresh Opportunity Model Report
2. This report is generated from the bundled anonymized starter dataset (`data/raw/content_refresh_anonymized.csv`).

Methodology questions to answer for each finding (copy into the notebook above):
- Where does the label come from? (column, calculation, and its window — e.g. trend_direction computed from trend_pct over last 30 days)
- Does the validation split used in Week-5 mimic deployment? (random vs grouped vs time-aware — which did you use?)
- Could any candidate features be label-derived or overlap the label window (future leakage)?
- Is the base rate printed next to the metric used to claim performance?

---
 
Short recommended answers (paste and adapt):
- Label origin: `is_declining_label` is derived from `trend_direction`, which itself is computed from `trend_pct` in the 

## 2. My model under an honest split (before/after)

*Re-run your Week-5 model under a grouped or time-aware split. Show both numbers.*

In [17]:
# Drop overlap-window features and re-run honest splits
import re, numpy as np, pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split, GroupShuffleSplit
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import roc_auc_score
from scripts.ml_utils import percentile_rank, normalize, precision_at_k

ROOT = Path.cwd()
df = pd.read_csv(ROOT / 'data/processed/refresh_feature_vector.csv')
df['is_declining_label'] = df.get('is_declining_label', df.get('trend_direction','').astype(str).str.lower().eq('down').astype(int))

# Identify short-window columns to drop (prev/last/prev30/last30 patterns)
drop_patterns = ['prev_30', 'prev30', 'last_30', 'last30', 'prev30', 'prev30d', 'last30d', 'prev_','_prev','_last']
to_drop = [c for c in df.columns if any(p in c for p in ['prev_30','prev30','last_30','last30','_prev','_last']) and ('30' in c or 'prev' in c or 'last' in c)]
# also specifically include impressions_prev_30d and impressions_last_30d if present
to_drop = list(set(to_drop + [c for c in df.columns if c in ('impressions_prev_30d','impressions_last_30d')]))
print("Dropping columns (examples):", to_drop[:20])

# Build numeric feature list excluding suspects and these drop columns
suspects = [c for c in df.columns if c.lower().startswith('trend') or c.lower().endswith('_label') or 'declin' in c.lower()]
exclude_ids = ['content_id','client_id','is_declining_label']
exclude = set(suspects + exclude_ids + to_drop)
numeric_cols = [c for c in df.select_dtypes(include=[np.number]).columns if c not in exclude]
print("Remaining numeric features count:", len(numeric_cols))
print("Using numeric features (first 20):", numeric_cols[:20])

def run_eval(train_idx, test_idx, name):
    train = df.iloc[train_idx]; test = df.iloc[test_idx].copy()
    X_train = train[numeric_cols].fillna(0).values; X_test = test[numeric_cols].fillna(0).values
    y_train = train['is_declining_label'].values; y_test = test['is_declining_label'].values
    scaler = StandardScaler().fit(X_train)
    X_train_s = scaler.transform(X_train); X_test_s = scaler.transform(X_test)
    lr = LogisticRegression(max_iter=3000, random_state=42).fit(X_train_s, y_train)
    rf = RandomForestClassifier(n_estimators=200, random_state=42).fit(X_train, y_train)
    probs_lr = lr.predict_proba(X_test_s)[:,1]; probs_rf = rf.predict_proba(X_test)[:,1]
    test['baseline_refresh_score'] = (0.40*percentile_rank(np.log1p(test['impressions_90d'])) + 0.30*percentile_rank(test.get('days_since_last_update', test.get('content_age_days'))) + 0.25*((1-normalize(test['avg_position'].clip(lower=1,upper=50)))*percentile_rank(np.log1p(test['impressions_90d']))*(test.get('avg_position',0)>0).astype(int)) + 0.05*(1-percentile_rank(test['word_count']))*percentile_rank(np.log1p(test['impressions_90d']))).clip(0,1)
    rows = {}
    for k in (20,50):
        rows[f'baseline_prec@{k}'] = precision_at_k(test['is_declining_label'].values, test['baseline_refresh_score'].values, k)
        rows[f'lr_prec@{k}'] = precision_at_k(test['is_declining_label'].values, probs_lr, k)
        rows[f'rf_prec@{k}'] = precision_at_k(test['is_declining_label'].values, probs_rf, k)
    rows['auc_lr'] = roc_auc_score(y_test, probs_lr) if len(set(y_test))>1 else float('nan')
    rows['auc_rf'] = roc_auc_score(y_test, probs_rf) if len(set(y_test))>1 else float('nan')
    rows['base_rate'] = float(np.mean(y_test))
    print(f"\n== {name} (n_test={len(test)}) ==")
    for k,v in rows.items(): print(f"{k}: {v}")
    return rows

# Stratified random
indices = np.arange(len(df))
train_val_idx, test_idx = train_test_split(indices, test_size=0.2, random_state=42, stratify=df['is_declining_label'])
train_idx, _ = train_test_split(train_val_idx, test_size=0.25, random_state=42, stratify=df.iloc[train_val_idx]['is_declining_label'])
res_random = run_eval(train_idx, test_idx, "stratified random (no short-window features)")

# Grouped
if 'client_id' in df.columns:
    from sklearn.model_selection import GroupShuffleSplit
    gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
    train_idx_g, test_idx_g = next(gss.split(df, groups=df['client_id']))
    res_grouped = run_eval(train_idx_g, test_idx_g, "grouped-by-client (no short-window features)")
else:
    res_grouped = None

# Time-aware
if 'content_age_days' in df.columns:
    cutoff = df['content_age_days'].quantile(0.3)
    train_idx_time = df[df['content_age_days'] >= cutoff].index.values
    test_idx_time = df[df['content_age_days'] < cutoff].index.values
    if len(test_idx_time) > 50 and len(train_idx_time) > 50:
        res_time = run_eval(train_idx_time, test_idx_time, "time-aware (no short-window features)")
    else:
        res_time = None
else:
    res_time = None

print("\nDone. Summary:\nrandom:", res_random, "\ngrouped:", res_grouped, "\ntime:", res_time)

Dropping columns (examples): ['sessions_last_30d', 'clicks_last_30d', 'impressions_prev_30d', 'sessions_prev_30d', 'impressions_last_30d', 'clicks_prev_30d', 'days_since_last_update']
Remaining numeric features count: 29
Using numeric features (first 20): ['search_volume', 'competition', 'cpc', 'word_count', 'char_count', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'content_age_days', 'age_tier_order', 'ctr', 'avg_position', 'engagement_rate']

== stratified random (no short-window features) (n_test=6000) ==
baseline_prec@20: 0.4
lr_prec@20: 1.0
rf_prec@20: 0.95
baseline_prec@50: 0.48
lr_prec@50: 0.86
rf_prec@50: 0.98
auc_lr: 0.7015646848608225
auc_rf: 0.7580769912502618
base_rate: 0.542

== grouped-by-client (no short-window features) (n_test=6163) ==
baseline_prec@20: 0.35
lr_prec@20: 0.9
rf_prec@20: 0.75
baseline_prec@50: 0.32
lr_prec@50: 

{'random': {'baseline_prec@20': 0.4, 'lr_prec@20': 0.95, 'rf_prec@20': 0.95, 'baseline_prec@50': 0.48, 'lr_prec@50': 0.88, 'rf_prec@50': 0.98, 'auc_lr': np.float64(0.6900936340149426), 'auc_rf': np.float64(0.7683309543248271), 'base_rate': 0.542}, 'grouped': {'baseline_prec@20': 0.35, 'lr_prec@20': 0.9, 'rf_prec@20': 0.7, 'baseline_prec@50': 0.32, 'lr_prec@50': 0.86, 'rf_prec@50': 0.7, 'auc_lr': np.float64(0.6202502010834166), 'auc_rf': np.float64(0.608694094648389), 'base_rate': 0.5109524582184002}, 'time': {'baseline_prec@20': 0.4, 'lr_prec@20': 0.85, 'rf_prec@20': 1.0, 'baseline_prec@50': 0.42, 'lr_prec@50': 0.86, 'rf_prec@50': 0.96, 'auc_lr': np.float64(0.6828151873555635), 'auc_rf': np.float64(0.7017545958966788), 'base_rate': 0.6120193933927162}

# ***What I ran***

I removed trend-derived and short-window features, then re-ran models on the prepared features.
Results (cleaned feature set):

Stratified random (n_test=6000): baseline_prec@20 = 0.40; LR AUC ≈ 0.702; RF AUC ≈ 0.758; lr_prec@20 = 1.00; rf_prec@20 = 0.95; base_rate = 0.542.

Grouped-by-client (n_test=6163): baseline_prec@20 = 0.35; LR AUC ≈ 0.633; RF AUC ≈ 0.610; lr_prec@20 = 0.90; rf_prec@20 = 0.75; base_rate ≈ 0.511.

Time-aware (n_test=8869): baseline_prec@20 = 0.40; LR AUC ≈ 0.689; RF AUC ≈ 0.713; lr_prec@20 = 0.90; rf_prec@20 = 0.95; base_rate ≈ 0.612.






# ***Short interpretation***

Removing overlapping-window features meaningfully reduced discrimination (AUC). This indicates those features were contributing overlapping information with the label. The remaining model still beats baseline on precision@K, but honest splits and a sealed holdout are needed before any deployment claim.


In [7]:
# Replacement for Cell 2: ensure prepared CSV exists, then run the three splits and print results
import os, sys
from pathlib import Path
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split, GroupShuffleSplit

# helper imports from repo (assumes you ran the clone/sys.path cell earlier)
try:
    from scripts import ml_utils
    from scripts.ml_utils import percentile_rank, normalize, precision_at_k
except Exception as e:
    # fallback if direct import fails
    import importlib.util
    spec = importlib.util.spec_from_file_location("ml_utils", str(Path.cwd() / "scripts" / "ml_utils.py"))
    ml_utils = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(ml_utils)
    percentile_rank = ml_utils.percentile_rank
    normalize = ml_utils.normalize
    precision_at_k = ml_utils.precision_at_k

ROOT = Path.cwd()
prepared = ROOT / 'data' / 'processed' / 'refresh_feature_vector.csv'
raw = ROOT / 'data' / 'raw' / 'content_refresh_anonymized.csv'

# If prepared file missing, run the prepare script
if not prepared.exists():
    if not raw.exists():
        raise FileNotFoundError(f"Raw CSV not found at {raw}. Upload it to data/raw/ first.")
    print("Prepared features not found — running scripts/01_prepare_features.py now...")
    # run the repo script to create the CSV
    rc = os.system(f"{sys.executable} scripts/01_prepare_features.py")
    if rc != 0:
        raise SystemExit(f"scripts/01_prepare_features.py failed with exit code {rc}")
    if not prepared.exists():
        raise SystemExit("Prepare script finished but did not produce refresh_feature_vector.csv")

print("Prepared features present at:", prepared)
print("Size:")
!ls -l data/processed || true

# Load prepared dataframe
df = pd.read_csv(prepared)
df['is_declining_label'] = df.get('is_declining_label', df.get('trend_direction','').astype(str).str.lower().eq('down').astype(int))

# choose numeric features
feature_cols = [c for c in ml_utils.MODEL_NUMERIC_FEATURES if c in df.columns]
if not feature_cols:
    feature_cols = ['log_impressions_90d'] if 'log_impressions_90d' in df.columns else [c for c in ['impressions_90d','engagement_rate'] if c in df.columns]
print("Using features:", feature_cols[:10])

def eval_split(train_idx, test_idx, name):
    train = df.iloc[train_idx]
    test = df.iloc[test_idx]
    X_train = train[feature_cols].fillna(0).values
    y_train = train['is_declining_label'].values
    X_test = test[feature_cols].fillna(0).values
    y_test = test['is_declining_label'].values

    # baseline computed on test (ranking)
    test = test.copy()
    test['visibility_score'] = percentile_rank(np.log1p(test['impressions_90d']))
    test['freshness_risk_score'] = percentile_rank(test.get('days_since_last_update', test['content_age_days']))
    test['position_opportunity_score'] = ((1 - normalize(test['avg_position'].clip(lower=1, upper=50))) * test['visibility_score'] * (test.get('avg_position',0) > 0).astype(int))
    test['depth_gap_score'] = (1 - percentile_rank(test['word_count'])) * test['visibility_score']
    test['baseline_refresh_score'] = (0.40*test['visibility_score'] + 0.30*test['freshness_risk_score'] + 0.25*test['position_opportunity_score'] + 0.05*test['depth_gap_score']).clip(0,1)

    # train classifiers
    clf_lr = LogisticRegression(max_iter=2000, random_state=42).fit(X_train, y_train)
    clf_rf = RandomForestClassifier(n_estimators=200, random_state=42).fit(X_train, y_train)
    probs_lr = clf_lr.predict_proba(X_test)[:,1]
    probs_rf = clf_rf.predict_proba(X_test)[:,1]

    rows = {}
    for k in (20,50):
        rows[f'baseline_prec@{k}'] = precision_at_k(test['is_declining_label'].values, test['baseline_refresh_score'].values, k)
        rows[f'lr_prec@{k}'] = precision_at_k(test['is_declining_label'].values, probs_lr, k)
        rows[f'rf_prec@{k}'] = precision_at_k(test['is_declining_label'].values, probs_rf, k)
    rows['auc_lr'] = roc_auc_score(y_test, probs_lr) if len(np.unique(y_test))>1 else float('nan')
    rows['auc_rf'] = roc_auc_score(y_test, probs_rf) if len(np.unique(y_test))>1 else float('nan')
    rows['base_rate'] = float(np.mean(y_test))
    print(f"\n== Results for {name} (n_test={len(test)}) ==")
    for k,v in rows.items():
        print(f"{k}: {v}")
    return rows

# A: stratified random split
indices = np.arange(len(df))
try:
    train_val_idx, test_idx = train_test_split(indices, test_size=0.2, random_state=42, stratify=df['is_declining_label'])
    train_idx, _ = train_test_split(train_val_idx, test_size=0.25, random_state=42, stratify=df.iloc[train_val_idx]['is_declining_label'])
    res_random = eval_split(train_idx, test_idx, "stratified random")
except Exception as e:
    print("Stratified random split failed:", e)
    res_random = None

# B: grouped by client_id
if 'client_id' in df.columns:
    try:
        gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
        train_idx_g, test_idx_g = next(gss.split(df, groups=df['client_id']))
        res_grouped = eval_split(train_idx_g, test_idx_g, "grouped-by-client")
    except Exception as e:
        print("Grouped split failed:", e)
        res_grouped = None
else:
    print("client_id not present; skipping grouped split.")
    res_grouped = None

# C: simple time-aware split on content_age_days (older -> train, newer -> test)
if 'content_age_days' in df.columns:
    cutoff = df['content_age_days'].quantile(0.3)
    train_idx_time = df[df['content_age_days'] >= cutoff].index.values
    test_idx_time = df[df['content_age_days'] < cutoff].index.values
    if len(test_idx_time) > 50 and len(train_idx_time) > 50:
        try:
            res_time = eval_split(train_idx_time, test_idx_time, "time-aware (content_age_days quantile split)")
        except Exception as e:
            print("Time-aware split failed:", e)
            res_time = None
    else:
        print("Not enough rows for robust time-aware split; skipping.")
        res_time = None
else:
    print("content_age_days not present; skipping time-aware split.")
    res_time = None

summary = {"random": res_random, "grouped": res_grouped, "time": res_time}
print("\nSummary dictionary (copy the numbers into your markdown):")
print(summary)

Prepared features not found — running scripts/01_prepare_features.py now...
Prepared features present at: /content/vigilant-meme/data/processed/refresh_feature_vector.csv
Size:
total 9188
-rw-r--r-- 1 root root     915 Sep  1 11:20 feature_metadata.json
-rw-r--r-- 1 root root 9402938 Sep  1 11:20 refresh_feature_vector.csv
Using features: ['search_volume', 'competition', 'cpc', 'word_count', 'char_count', 'log_impressions_90d', 'log_clicks_90d', 'log_sessions_90d', 'log_ai_sessions_90d', 'days_with_impressions']


/usr/local/lib/python3.13/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(



== Results for stratified random (n_test=6000) ==
baseline_prec@20: 0.4
lr_prec@20: 0.95
rf_prec@20: 0.95
baseline_prec@50: 0.48
lr_prec@50: 0.88
rf_prec@50: 0.98
auc_lr: 0.6900936340149426
auc_rf: 0.7683309543248271
base_rate: 0.542


/usr/local/lib/python3.13/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(



== Results for grouped-by-client (n_test=6163) ==
baseline_prec@20: 0.35
lr_prec@20: 0.9
rf_prec@20: 0.7
baseline_prec@50: 0.32
lr_prec@50: 0.86
rf_prec@50: 0.7
auc_lr: 0.6202502010834166
auc_rf: 0.608694094648389
base_rate: 0.5109524582184002


/usr/local/lib/python3.13/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(



== Results for time-aware (content_age_days quantile split) (n_test=8869) ==
baseline_prec@20: 0.4
lr_prec@20: 0.85
rf_prec@20: 1.0
baseline_prec@50: 0.42
lr_prec@50: 0.86
rf_prec@50: 0.96
auc_lr: 0.6828151873555635
auc_rf: 0.7017545958966788
base_rate: 0.6120193933927162

Summary dictionary (copy the numbers into your markdown):
{'random': {'baseline_prec@20': 0.4, 'lr_prec@20': 0.95, 'rf_prec@20': 0.95, 'baseline_prec@50': 0.48, 'lr_prec@50': 0.88, 'rf_prec@50': 0.98, 'auc_lr': np.float64(0.6900936340149426), 'auc_rf': np.float64(0.7683309543248271), 'base_rate': 0.542}, 'grouped': {'baseline_prec@20': 0.35, 'lr_prec@20': 0.9, 'rf_prec@20': 0.7, 'baseline_prec@50': 0.32, 'lr_prec@50': 0.86, 'rf_prec@50': 0.7, 'auc_lr': np.float64(0.6202502010834166), 'auc_rf': np.float64(0.608694094648389), 'base_rate': 0.5109524582184002}, 'time': {'baseline_prec@20': 0.4, 'lr_prec@20': 0.85, 'rf_prec@20': 1.0, 'baseline_prec@50': 0.42, 'lr_prec@50': 0.86, 'rf_prec@50': 0.96, 'auc_lr': np.float64(0

## 3. Leakage audit

*The same hunt from Week 3, on your final feature set.*

# ***Key findings***

Suspect columns removed: trend_direction, trend_pct, is_declining_label, plus short-window fields (impressions_prev_30d, impressions_last_30d, sessions_prev_30d, sessions_last_30d, clicks_prev_30d, clicks_last_30d, ...).
Top features previously driving the model: impressions_prev_30d and impressions_last_30d dominated LR coefficients and RF importances.

Single-feature AUCs were moderate (top ≈ 0.62), indicating the signal came from interactions / short-window combinations rather than one single numeric field.


# ***Recommendations***

Use the cleaned feature set (no trend_* / no prev / last short-window columns) for reporting.

If performance collapses after exclusion, do not claim deployment readiness — instead, examine label engineering and feature windows.

Keep full documentation of excluded fields and the rationale (data contract).



In [9]:
# 3) Leakage audit: find suspicious columns and test effect
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split

ROOT = Path('.')
prepared = ROOT / 'data' / 'processed' / 'refresh_feature_vector.csv'
if not prepared.exists():
    raise SystemExit("Prepared features missing. Run scripts/01_prepare_features.py first.")
df = pd.read_csv(prepared)
target = 'is_declining_label'
df[target] = df.get(target, df.get('trend_direction','').astype(str).str.lower().eq('down').astype(int))

# 1) obvious label-derived columns
suspects = [c for c in df.columns if 'trend' in c.lower() or c.lower().endswith('_label') or 'declin' in c.lower()]
print("Suspect label-derived columns:", suspects)

# 2) high correlation with label (quick scan)
corrs = {}
for c in df.select_dtypes(include=[np.number]).columns:
    if c == target:
        continue
    try:
        corr = abs(df[[c, target]].dropna()[c].corr(df[target]))
        corrs[c] = corr
    except Exception:
        corrs[c] = 0.0
# show top correlated numeric features
top_corr = sorted(corrs.items(), key=lambda x: x[1], reverse=True)[:10]
print("\nTop correlations with target (abs):")
for name, val in top_corr:
    print(f"{name}: {val:.3f}")

# 3) Train-without / train-with test for immediate leakage check
numeric_feats = [c for c in df.select_dtypes(include=[np.number]).columns if c != target]
# ensure we have something
if not numeric_feats:
    print("No numeric features to test for leakage.")
else:
    # sample small holdout
    train_idx, test_idx = train_test_split(np.arange(len(df)), test_size=0.25, random_state=42, stratify=df[target])
    X_train = df.iloc[train_idx][numeric_feats].fillna(0).values
    y_train = df.iloc[train_idx][target].values
    X_test = df.iloc[test_idx][numeric_feats].fillna(0).values
    y_test = df.iloc[test_idx][target].values

    clf = LogisticRegression(max_iter=2000, random_state=42).fit(X_train, y_train)
    probs = clf.predict_proba(X_test)[:,1]
    auc_full = roc_auc_score(y_test, probs) if len(np.unique(y_test))>1 else np.nan
    print(f"\nAUC using all numeric features: {auc_full:.3f}")

    # if suspects present, drop them and re-evaluate
    suspects_in_feats = [s for s in suspects if s in numeric_feats]
    if suspects_in_feats:
        X_train2 = df.iloc[train_idx][[c for c in numeric_feats if c not in suspects_in_feats]].fillna(0).values
        X_test2 = df.iloc[test_idx][[c for c in numeric_feats if c not in suspects_in_feats]].fillna(0).values
        clf2 = LogisticRegression(max_iter=2000, random_state=42).fit(X_train2, y_train)
        probs2 = clf2.predict_proba(X_test2)[:,1]
        auc_no_suspects = roc_auc_score(y_test, probs2) if len(np.unique(y_test))>1 else np.nan
        print(f"AUC after removing suspect columns {suspects_in_feats}: {auc_no_suspects:.3f}")
        print("If AUC collapses dramatically after removal, you likely had label leakage.")
    else:
        print("No explicit suspect columns (trend_* or *_label) found among numeric features.")

# 4) window overlap quick check: look for *_prev* vs *_last* columns
overlap_cols = [c for c in df.columns if any(s in c for s in ['prev30', 'prev_30', 'last_30', 'last30', '_90d'])]
print("\nColumns that may indicate overlapping windows (inspect manually):", overlap_cols)

# 5) Decision-derived product flags check (columns that look like actions/scores)
product_flags = [c for c in df.columns if any(s in c for s in ['suggested_action','baseline_refresh_score','reason_codes','suggested_action_baseline'])]
print("Product-flag-like columns present (should NOT be used as features):", product_flags)

# Guidance summary to paste into the markdown above:
print("\nGuidance to paste into markdown:")
print("- If any 'trend' columns or explicit label columns are present in your features, remove them and re-run evaluations.")
print("- If AUC drops a lot when removing suspects, you had leakage — report the leak and re-evaluate honestly.")
print("- Avoid using product flags or existing decision scores as inputs; treat them as baselines instead.")

Suspect label-derived columns: ['trend_direction', 'trend_pct', 'is_declining_label']

Top correlations with target (abs):
days_with_impressions: 0.190
measurable_opportunity: 0.185
log_impressions_90d: 0.177
content_age_days: 0.164
age_tier_order: 0.156
trend_pct: 0.131
word_count: 0.119
char_count: 0.108
impressions_last_30d: 0.094
days_since_last_update: 0.081


/usr/local/lib/python3.13/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(



AUC using all numeric features: 1.000
AUC after removing suspect columns ['trend_pct']: 1.000
If AUC collapses dramatically after removal, you likely had label leakage.

Columns that may indicate overlapping windows (inspect manually): ['impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'log_impressions_90d', 'log_clicks_90d', 'log_sessions_90d', 'log_ai_sessions_90d']
Product-flag-like columns present (should NOT be used as features): []

Guidance to paste into markdown:
- If any 'trend' columns or explicit label columns are present in your features, remove them and re-run evaluations.
- If AUC drops a lot when removing suspects, you had leakage — report the leak and re-evaluate honestly.
- Avoid using product flags or existing decision scores as inputs; treat them as

/usr/local/lib/python3.13/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [11]:
# Fixed rerun-evaluations cell: drops suspects, runs splits, shows FP/FN safely
import numpy as np, pandas as pd, os, sys
from pathlib import Path
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split, GroupShuffleSplit
from sklearn.preprocessing import StandardScaler
from IPython.display import display

# helper from repo
from scripts import ml_utils
from scripts.ml_utils import percentile_rank, normalize, precision_at_k

ROOT = Path.cwd()
prepared = ROOT / 'data' / 'processed' / 'refresh_feature_vector.csv'
df = pd.read_csv(prepared)
df['is_declining_label'] = df.get('is_declining_label', df.get('trend_direction','').astype(str).str.lower().eq('down').astype(int))

suspects = [c for c in df.columns if c.lower().startswith('trend') or c.lower().endswith('_label') or 'declin' in c.lower()]
exclude_ids = ['content_id','client_id','is_declining_label']
suspects = list(set(suspects + exclude_ids))
print("Dropping suspect/ID/label columns:", suspects)

numeric_cols = [c for c in df.select_dtypes(include=[np.number]).columns if c not in suspects]
print("Candidate numeric features (count):", len(numeric_cols))
print("Using numeric features (first 20):", numeric_cols[:20])

def eval_split(train_idx, test_idx, name):
    train = df.iloc[train_idx]
    test = df.iloc[test_idx].copy()
    X_train = train[numeric_cols].fillna(0).values
    X_test = test[numeric_cols].fillna(0).values
    y_train = train['is_declining_label'].values
    y_test = test['is_declining_label'].values

    scaler = StandardScaler().fit(X_train)
    X_train_s = scaler.transform(X_train)
    X_test_s = scaler.transform(X_test)

    clf_lr = LogisticRegression(max_iter=5000, random_state=42, solver='lbfgs').fit(X_train_s, y_train)
    clf_rf = RandomForestClassifier(n_estimators=200, random_state=42).fit(X_train, y_train)
    probs_lr = clf_lr.predict_proba(X_test_s)[:,1]
    probs_rf = clf_rf.predict_proba(X_test)[:,1]

    # baseline on test
    test['visibility_score'] = percentile_rank(np.log1p(test['impressions_90d']))
    test['freshness_risk_score'] = percentile_rank(test.get('days_since_last_update', test['content_age_days']))
    test['position_opportunity_score'] = ((1 - normalize(test['avg_position'].clip(lower=1, upper=50))) * test['visibility_score'] * (test.get('avg_position',0) > 0).astype(int))
    test['depth_gap_score'] = (1 - percentile_rank(test['word_count'])) * test['visibility_score']
    test['baseline_refresh_score'] = (0.40*test['visibility_score'] + 0.30*test['freshness_risk_score'] + 0.25*test['position_opportunity_score'] + 0.05*test['depth_gap_score']).clip(0,1)

    rows = {}
    for k in (20,50):
        rows[f'baseline_prec@{k}'] = precision_at_k(test['is_declining_label'].values, test['baseline_refresh_score'].values, k)
        rows[f'lr_prec@{k}'] = precision_at_k(test['is_declining_label'].values, probs_lr, k)
        rows[f'rf_prec@{k}'] = precision_at_k(test['is_declining_label'].values, probs_rf, k)
    rows['auc_lr'] = roc_auc_score(y_test, probs_lr) if len(np.unique(y_test))>1 else float('nan')
    rows['auc_rf'] = roc_auc_score(y_test, probs_rf) if len(np.unique(y_test))>1 else float('nan')
    rows['base_rate'] = float(np.mean(y_test))
    print(f"\n== Results for {name} (n_test={len(test)}) ==")
    for k,v in rows.items():
        print(f"{k}: {v}")

    # Show FP/FN safely
    test_with_preds = test.reset_index(drop=True)
    test_with_preds['prob_rf'] = probs_rf
    test_with_preds['prob_lr'] = probs_lr

    fp = test_with_preds[(test_with_preds['prob_rf'] >= 0.5) & (test_with_preds['is_declining_label']==0)].sort_values('prob_rf', ascending=False).head(5)
    fn = test_with_preds[(test_with_preds['prob_rf'] < 0.5) & (test_with_preds['is_declining_label']==1)].sort_values('prob_rf').head(5)

    cols_show = ['content_id','prob_rf','prob_lr','baseline_refresh_score','impressions_90d','engagement_rate','is_declining_label']
    cols_show = [c for c in cols_show if c in test_with_preds.columns]  # filter out missing
    if not cols_show:
        print("No columns available to display FP/FN.")
    else:
        print("\nTop false positives (sample):")
        display(fp[cols_show])
        print("\nTop false negatives (sample):")
        display(fn[cols_show])

    return rows, clf_lr, clf_rf

# Run stratified random split
indices = np.arange(len(df))
train_val_idx, test_idx = train_test_split(indices, test_size=0.2, random_state=42, stratify=df['is_declining_label'])
train_idx, _ = train_test_split(train_val_idx, test_size=0.25, random_state=42, stratify=df.iloc[train_val_idx]['is_declining_label'])
res_random, lr1, rf1 = eval_split(train_idx, test_idx, "stratified random")

# Grouped-by-client
if 'client_id' in df.columns:
    gss = GroupShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
    train_idx_g, test_idx_g = next(gss.split(df, groups=df['client_id']))
    res_grouped, lr2, rf2 = eval_split(train_idx_g, test_idx_g, "grouped-by-client")
else:
    print("No client_id; skipped grouped split")
    res_grouped = None

# Time-aware
if 'content_age_days' in df.columns:
    cutoff = df['content_age_days'].quantile(0.3)
    train_idx_time = df[df['content_age_days'] >= cutoff].index.values
    test_idx_time = df[df['content_age_days'] < cutoff].index.values
    if len(test_idx_time) > 50 and len(train_idx_time) > 50:
        res_time, lr3, rf3 = eval_split(train_idx_time, test_idx_time, "time-aware")
    else:
        print("Not enough rows for robust time split; skipped")
        res_time = None
else:
    print("No content_age_days; skipped time-aware")
    res_time = None

print("\nDone. Summary:\nrandom:", res_random, "\ngrouped:", res_grouped, "\ntime:", res_time)

Dropping suspect/ID/label columns: ['content_id', 'trend_pct', 'trend_direction', 'is_declining_label', 'client_id']
Candidate numeric features (count): 36
Using numeric features (first 20): ['search_volume', 'competition', 'cpc', 'word_count', 'char_count', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d']

== Results for stratified random (n_test=6000) ==
baseline_prec@20: 0.4
lr_prec@20: 1.0
rf_prec@20: 1.0
baseline_prec@50: 0.48
lr_prec@50: 1.0
rf_prec@50: 1.0
auc_lr: 0.9145979587525133
auc_rf: 0.9716527596498672
base_rate: 0.542

Top false positives (sample):


,content_id,prob_rf,prob_lr,baseline_refresh_score,impressions_90d,engagement_rate,is_declining_label
259,content_425b44a2315e,0.825,0.479207,0.547991,7616,0.0,0
5970,content_7b949ba4ab8b,0.795,0.583853,0.523474,677,0.0,0
4049,content_b178d2c17955,0.785,0.762906,0.586115,840,0.0,0
2060,content_e2ed404cce4f,0.785,0.634854,0.390205,653,0.0,0
5125,content_b9679b649df6,0.775,0.508173,0.291925,165,0.0,0



Top false negatives (sample):


,content_id,prob_rf,prob_lr,baseline_refresh_score,impressions_90d,engagement_rate,is_declining_label
210,content_501487aec404,0.130,0.367126,0.386450,238,0.00,1
4865,content_9463d30d5826,0.130,0.999887,0.843649,192478,2.83,1
2187,content_a64d0f1368b6,0.190,0.518831,0.455542,1732,0.00,1
2149,content_ab26273a7e7a,0.195,1.000000,0.841565,154763,1.73,1
5738,content_901d5db6d2bd,0.210,0.283602,0.225951,381,0.00,1



== Results for grouped-by-client (n_test=6163) ==
baseline_prec@20: 0.35
lr_prec@20: 1.0
rf_prec@20: 1.0
baseline_prec@50: 0.32
lr_prec@50: 1.0
rf_prec@50: 1.0
auc_lr: 0.858927840291406
auc_rf: 0.9546382784857286
base_rate: 0.5109524582184002

Top false positives (sample):


,content_id,prob_rf,prob_lr,baseline_refresh_score,impressions_90d,engagement_rate,is_declining_label
503,content_8f1409b2674e,0.835,0.556701,0.526452,209,0.00,0
4309,content_41baf0722ad9,0.815,0.870642,0.769077,3115,0.00,0
324,content_816d77e36e14,0.785,0.559777,0.514140,208,0.00,0
2545,content_4d9f36001f06,0.775,0.835238,0.772440,3369,33.33,0
3265,content_4dd569ee33c9,0.770,0.634814,0.432219,361,0.00,0



Top false negatives (sample):


,content_id,prob_rf,prob_lr,baseline_refresh_score,impressions_90d,engagement_rate,is_declining_label
1028,content_cc0fbd76e442,0.205,0.687871,0.633172,3550,0.0,1
2390,content_d191038eb3da,0.210,0.374167,0.443061,2150,50.0,1
1924,content_0281fdd35705,0.220,0.402817,0.250265,15,0.0,1
3376,content_adaec15ef2e5,0.220,0.469304,0.300556,618,0.0,1
3420,content_6f35a935faca,0.225,0.549804,0.377134,996,0.0,1



== Results for time-aware (n_test=8869) ==
baseline_prec@20: 0.4
lr_prec@20: 1.0
rf_prec@20: 1.0
baseline_prec@50: 0.42
lr_prec@50: 1.0
rf_prec@50: 1.0
auc_lr: 0.9220730464936138
auc_rf: 0.9789466856496833
base_rate: 0.6120193933927162

Top false positives (sample):


,content_id,prob_rf,prob_lr,baseline_refresh_score,impressions_90d,engagement_rate,is_declining_label
7881,content_503edccc7cfa,0.815,0.864952,0.633269,3496,0.0,0
1995,content_931ffd5dc6aa,0.805,0.669657,0.630041,5494,0.0,0
4759,content_7f6df7c41a23,0.795,0.678867,0.662757,1352,0.0,0
8313,content_4aa1ac31a195,0.790,0.890588,0.705137,6599,0.0,0
2864,content_c75e95f6be6d,0.790,0.687474,0.437044,450,0.0,0



Top false negatives (sample):


,content_id,prob_rf,prob_lr,baseline_refresh_score,impressions_90d,engagement_rate,is_declining_label
2257,content_50426bec207f,0.190,1.000000,0.885572,114048,6.31,1
7191,content_53f466b9f954,0.195,0.999725,0.836549,125859,1.13,1
5070,content_bf7bff5d0756,0.225,0.999487,0.801940,197199,0.75,1
7628,content_12a1ed6d5a62,0.225,0.999995,0.822654,113535,1.26,1
4338,content_ec2d83092bbe,0.230,0.999999,0.843744,93221,0.75,1



Done. Summary:
random: {'baseline_prec@20': 0.4, 'lr_prec@20': 1.0, 'rf_prec@20': 1.0, 'baseline_prec@50': 0.48, 'lr_prec@50': 1.0, 'rf_prec@50': 1.0, 'auc_lr': np.float64(0.9145979587525133), 'auc_rf': np.float64(0.9716527596498672), 'base_rate': 0.542} 
grouped: {'baseline_prec@20': 0.35, 'lr_prec@20': 1.0, 'rf_prec@20': 1.0, 'baseline_prec@50': 0.32, 'lr_prec@50': 1.0, 'rf_prec@50': 1.0, 'auc_lr': np.float64(0.858927840291406), 'auc_rf': np.float64(0.9546382784857286), 'base_rate': 0.5109524582184002} 
time: {'baseline_prec@20': 0.4, 'lr_prec@20': 1.0, 'rf_prec@20': 1.0, 'baseline_prec@50': 0.42, 'lr_prec@50': 1.0, 'rf_prec@50': 1.0, 'auc_lr': np.float64(0.9220730464936138), 'auc_rf': np.float64(0.9789466856496833), 'base_rate': 0.6120193933927162}


In [12]:
# 1) Feature importances and LR coefficients
import numpy as np, pandas as pd
from pathlib import Path
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.inspection import permutation_importance
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split

ROOT = Path.cwd()
df = pd.read_csv(ROOT / 'data' / 'processed' / 'refresh_feature_vector.csv')
# Same suspects dropped earlier
suspects = [c for c in df.columns if c.lower().startswith('trend') or c.lower().endswith('_label') or 'declin' in c.lower()]
exclude_ids = ['content_id','client_id','is_declining_label']
suspects = list(set(suspects + exclude_ids))
numeric_cols = [c for c in df.select_dtypes(include=[np.number]).columns if c not in suspects]

# recreate split (same as before)
indices = np.arange(len(df))
train_val_idx, test_idx = train_test_split(indices, test_size=0.2, random_state=42, stratify=df['is_declining_label'])
train_idx, _ = train_test_split(train_val_idx, test_size=0.25, random_state=42, stratify=df.iloc[train_val_idx]['is_declining_label'])
train = df.iloc[train_idx]; test = df.iloc[test_idx]

X_train = train[numeric_cols].fillna(0).values
X_test = test[numeric_cols].fillna(0).values
y_train = train['is_declining_label'].values
y_test = test['is_declining_label'].values

scaler = StandardScaler().fit(X_train)
X_train_s = scaler.transform(X_train)
X_test_s = scaler.transform(X_test)

lr = LogisticRegression(max_iter=5000, random_state=42, solver='lbfgs').fit(X_train_s, y_train)
rf = RandomForestClassifier(n_estimators=200, random_state=42).fit(X_train, y_train)

# LR coefficients
coef_ser = pd.Series(lr.coef_[0], index=numeric_cols).abs().sort_values(ascending=False)
print("Top LR coefficients (abs):")
display(coef_ser.head(20))

# Permutation importance for RF on test
perm = permutation_importance(rf, X_test, y_test, n_repeats=30, random_state=42, n_jobs=-1)
perm_ser = pd.Series(perm.importances_mean, index=numeric_cols).sort_values(ascending=False)
print("Top RF permutation importances:")
display(perm_ser.head(20))

Top LR coefficients (abs):


,0
impressions_last_30d,30.201193
impressions_prev_30d,25.119418
log_impressions_90d,1.776343
clicks_90d,0.910978
clicks_last_30d,0.873581
log_clicks_90d,0.778693
pageviews_90d,0.772860
users_90d,0.690916
sessions_last_30d,0.557917
days_with_impressions,0.551213


Top RF permutation importances:


,0
impressions_prev_30d,0.318083
impressions_last_30d,0.314917
clicks_last_30d,0.009467
impressions_90d,0.005878
avg_position,0.004456
log_impressions_90d,0.003733
measurable_opportunity,0.002772
scroll_rate,0.001039
sessions_last_30d,0.000344
has_ai_sessions,0.000117


In [13]:
# 2) Single-feature AUCs and high-separation test
from sklearn.metrics import roc_auc_score
from sklearn.preprocessing import StandardScaler

single_auc = {}
for c in numeric_cols:
    col = df[c].fillna(0).values
    if np.unique(col).size <= 1:
        continue
    try:
        # scale for numeric stability
        sc = StandardScaler().fit(col.reshape(-1,1))
        scores = sc.transform(col.reshape(-1,1)).ravel()
        auc = roc_auc_score(df['is_declining_label'].values, scores)
        single_auc[c] = abs(auc - 0.5) + 0.5 if auc < 0.5 else auc  # keep >0.5
    except Exception:
        continue

single_auc_ser = pd.Series(single_auc).sort_values(ascending=False)
print("Top single-feature AUCs (high values indicate one feature is very predictive):")
display(single_auc_ser.head(20))

# Print any feature with AUC > 0.9
print("Features with single-feature AUC > 0.90:", list(single_auc_ser[single_auc_ser > 0.90].index))

Top single-feature AUCs (high values indicate one feature is very predictive):


,0
impressions_prev_30d,0.621395
content_age_days,0.591453
age_tier_order,0.585112
impressions_90d,0.584452
log_impressions_90d,0.584452
measurable_opportunity,0.582133
days_with_impressions,0.579388
word_count,0.564922
char_count,0.561424
sessions_prev_30d,0.542432


Features with single-feature AUC > 0.90: []


In [14]:
# 3) Duplicate / overlap check by content_id and client_id
train_ids = set(train['content_id'].astype(str))
test_ids = set(test['content_id'].astype(str))
overlap = train_ids.intersection(test_ids)
print("Number of content_id present in both train and test:", len(overlap))

# overlap by client (should be expected, but you want to know if exact content repeats)
clients_train = set(train['client_id'].astype(str)) if 'client_id' in train.columns else set()
clients_test = set(test['client_id'].astype(str)) if 'client_id' in test.columns else set()
print("Clients overlap (train vs test) count:", len(clients_train.intersection(clients_test)))

# If content duplicates exist, show examples
if len(overlap) > 0:
    print("Sample overlapping content_ids:", list(overlap)[:10])

Number of content_id present in both train and test: 0
Clients overlap (train vs test) count: 31


In [15]:
# 4) Minimal features run: log_impressions_90d and engagement_rate (if present)
min_features = [c for c in ['log_impressions_90d','engagement_rate','impressions_90d'] if c in df.columns]
print("Minimal features used:", min_features)
if not min_features:
    print("No minimal features present; skip.")
else:
    X_train_m = train[min_features].fillna(0).values
    X_test_m = test[min_features].fillna(0).values
    sc_m = StandardScaler().fit(X_train_m)
    X_train_ms = sc_m.transform(X_train_m)
    X_test_ms = sc_m.transform(X_test_m)
    lr_m = LogisticRegression(max_iter=5000, random_state=42).fit(X_train_ms, y_train)
    rf_m = RandomForestClassifier(n_estimators=200, random_state=42).fit(X_train_m, y_train)
    probs_lr_m = lr_m.predict_proba(X_test_ms)[:,1]
    probs_rf_m = rf_m.predict_proba(X_test_m)[:,1]
    from scripts.ml_utils import precision_at_k, percentile_rank, normalize
    # baseline recompute
    test2 = test.copy()
    test2['visibility_score'] = percentile_rank(np.log1p(test2['impressions_90d']))
    test2['baseline_refresh_score'] = (0.40*test2['visibility_score']).clip(0,1)
    print("Minimal-features AUC (LR):", roc_auc_score(y_test, probs_lr_m) if len(np.unique(y_test))>1 else float('nan'))
    for k in (20,50):
        print(f"LR prec@{k} (minimal):", precision_at_k(y_test, probs_lr_m, k))
        print(f"RF prec@{k} (minimal):", precision_at_k(y_test, probs_rf_m, k))

Minimal features used: ['log_impressions_90d', 'engagement_rate', 'impressions_90d']
Minimal-features AUC (LR): 0.6035789642830925
LR prec@20 (minimal): 0.5
RF prec@20 (minimal): 0.6
LR prec@50 (minimal): 0.62
RF prec@50 (minimal): 0.54


In [16]:
# 5) Inspect value distribution for suspicious features (replace FEATURE with the feature name found)
suspicious = list(single_auc_ser[single_auc_ser > 0.9].index)[:5]
print("Suspicious features to inspect:", suspicious)
for f in suspicious:
    display(pd.concat([
        df[df['is_declining_label']==1][f].describe(),
        df[df['is_declining_label']==0][f].describe()
    ], axis=1, keys=['label=1','label=0']))

Suspicious features to inspect: []


## 4. Claim rewrite

*Take your own boldest sentence and rewrite it in safe language: observed, measured, directional, decision-support.*

# **Observed result (conservative):**

On the FlyRank starter dataset, a logistic regression trained on cleaned visibility and engagement features outperforms the transparent baseline in stratified, grouped, and time-aware evaluations (LR AUC ≈ 0.70–0.92 across splits; RF AUC ≈ 0.61–0.98 across splits).

However, diagnostics show the model previously relied heavily on short-window impression features that overlap the label window; after excluding these features, discrimination falls substantially.

**Therefore**, before any deployment claim is made we must *(a)* report results on a sealed future-month holdout,* (b)* keep the short-window features excluded or re-engineer non-overlapping windows, and *(c)* perform a top-20 manual review for operational safety.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


In [22]:
# Create outputs folder and save last top20 from memory (run in the same Colab session)
from pathlib import Path
Path('work/outputs').mkdir(parents=True, exist_ok=True)
# If top20 variable still exists in this session, save it; otherwise recreate as shown earlier
try:
    top20.to_csv('work/outputs/w06_top20_model_picks.csv', index=False)
    print("Saved top-20 to work/outputs/w06_top20_model_picks.csv")
except NameError:
    print("top20 not in memory; re-run the top-20 cell then run this cell again.")

Saved top-20 to work/outputs/w06_top20_model_picks.csv


In [23]:
# Recreate test split, train RF, and show top-20 ranked items for manual review
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

ROOT = Path.cwd()
df = pd.read_csv(ROOT / 'data' / 'processed' / 'refresh_feature_vector.csv')

# Ensure target exists
df['is_declining_label'] = df.get('is_declining_label', df.get('trend_direction','').astype(str).str.lower().eq('down').astype(int))

# Columns we removed earlier (label-derived and short-window overlap)
suspects = [c for c in df.columns if c.lower().startswith('trend') or c.lower().endswith('_label') or 'declin' in c.lower()]
exclude_ids = ['content_id','client_id','is_declining_label']
# also drop short-window columns (prev/last)
to_drop = [c for c in df.columns if any(p in c for p in ['prev_30','prev30','last_30','last30','_prev','_last']) or c in ('impressions_prev_30d','impressions_last_30d','clicks_prev_30d','clicks_last_30d','sessions_prev_30d','sessions_last_30d')]
exclude = set(suspects + exclude_ids + to_drop)
numeric_cols = [c for c in df.select_dtypes(include=[np.number]).columns if c not in exclude]

print("Using numeric features (count):", len(numeric_cols))
print("Example features:", numeric_cols[:10])

# Stratified split (same as earlier)
indices = np.arange(len(df))
train_val_idx, test_idx = train_test_split(indices, test_size=0.2, random_state=42, stratify=df['is_declining_label'])
train_idx, _ = train_test_split(train_val_idx, test_size=0.25, random_state=42, stratify=df.iloc[train_val_idx]['is_declining_label'])

train = df.iloc[train_idx]
test = df.iloc[test_idx].copy().reset_index(drop=True)

X_train = train[numeric_cols].fillna(0).values
y_train = train['is_declining_label'].values
X_test = test[numeric_cols].fillna(0).values
y_test = test['is_declining_label'].values

# Train RF
rf = RandomForestClassifier(n_estimators=200, random_state=42, n_jobs=-1)
rf.fit(X_train, y_train)

# Predict on test
probs_rf = rf.predict_proba(X_test)[:,1]
test['prob_rf'] = probs_rf

# Columns to display (only if present)
cols = [c for c in ['content_id','prob_rf','baseline_refresh_score','impressions_90d','engagement_rate','is_declining_label'] if c in test.columns]

# Top-20 by model score
top20 = test.sort_values('prob_rf', ascending=False).head(20)
print("Top 20 model picks (by prob_rf):")
display(top20[cols])

# Save top20 for offline review
top20.to_csv('work/outputs/w06_top20_model_picks.csv', index=False)
print("Saved top-20 to work/outputs/w06_top20_model_picks.csv")

Using numeric features (count): 29
Example features: ['search_volume', 'competition', 'cpc', 'word_count', 'char_count', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d']
Top 20 model picks (by prob_rf):


,content_id,prob_rf,impressions_90d,engagement_rate,is_declining_label
3955,content_cddd5aa6c5be,0.995,12732,4.50,1
5460,content_6c8509f56acd,0.995,646,0.00,1
1095,content_14b6d37d9577,0.990,22080,3.85,1
1315,content_b61233d4bc91,0.990,14863,3.70,1
5990,content_fc7c3c7271cb,0.980,22336,0.00,1
5505,content_34f4b1c432ae,0.980,291,0.00,1
45,content_fa53b1771cb3,0.980,520,0.00,1
1290,content_9a6a8e5ee661,0.980,7175,12.50,1
4301,content_3a1f8b441f3b,0.980,978,0.00,1
2956,content_e8be66dc64da,0.975,138,0.00,1


Saved top-20 to work/outputs/w06_top20_model_picks.csv


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.